### PR TITLE
Proper support for namespaces and other enhancement/fixes

### LINKED ISSUE #17409: Godot ignores namespace information when instantiating C# classes
**Godot version:**
`master` / dad47d8876de45132226985dbbd079e060f6facc

**OS/device including version:**
_Manjaro 17.1_

**Issue description:**
Godot apparently doesn't take the namespace information into account when instantiating a class that is assigned to a node.

**Steps to reproduce:**
 * Create `Namespace2/MyClass.cs` and assign it to a node.
 * Override `Namespace2.MyClass._Ready()` and add ```GD.Print("MyClass 2")```.
 * Run the project and verify `MyClass 2` is printed in the console.
 * Create `Namespace1/MyClass.cs` but do _not_ assign it to anything.
 * Override `Namespace1.MyClass._Ready()` and add ```GD.Print("MyClass 1")```.
 * Run the project again, and verify `MyClass 1` is printed instead of `MyClass 2`.

### PR DESCRIPTION
**Pull request by neikeq**
- Fix internal assembly load from (it wasn't actually loading from the specified path) and make sure we load API assemblies from `res://.mono/assemblies/` instead of other locations.
- Support globs in csproj includes.
- Do not generate API project GUIDs randomly.
- Fix msvc warnings in mono module.
- Proper namespace support: Parse C# script namespace and class. Fixes #17409

The parser to extract the class name and namespace is an improved version of the one uploaded by @reduz (4613cb7874a7a6beb3195bebf06d7ecfbd2454c3), with a few modifications and fixes.
If this turns not to be a reliable solution, we can always switch to Microsoft.CodeAnalysis.CSharp.

**DNM:** This PR adds a tiny dependency for the editor (https://github.com/dazinator/DotNet.Glob), and I'm still trying to figure out where the license info should go.

### CODE REVIEW

**Comment by exts** [comment]
Building on windows 10 failed for me:

```
C:\Program Files\Mono\lib\mono\msbuild\15.0\bin\NuGet.targets(104,5): error : Unable to load the service index for sour
ce https://api.nuget.org/v3/index.json. [D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTool
s.csproj]
C:\Program Files\Mono\lib\mono\msbuild\15.0\bin\NuGet.targets(104,5): error :   An error occurred while sending the req
uest [D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTools.csproj]
C:\Program Files\Mono\lib\mono\msbuild\15.0\bin\NuGet.targets(104,5): error :   Error: TrustFailure (The authentication
 or decryption has failed.) [D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTools.csproj]
C:\Program Files\Mono\lib\mono\msbuild\15.0\bin\NuGet.targets(104,5): error :   The authentication or decryption has fa
iled. [D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTools.csproj]
C:\Program Files\Mono\lib\mono\msbuild\15.0\bin\NuGet.targets(104,5): error :   Invalid certificate received from serve
r. Error code: 0xffffffff800b0109 [D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTools.cspr
oj]
Done Building Project "D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTools.csproj" (Restore
 target(s)) -- FAILED.

Done Building Project "D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTools.sln" (Restore;Bu
ild target(s)) -- FAILED.


Build FAILED.

"D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTools.sln" (Restore;Build target) (1) ->
"D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTools.csproj" (Restore target) (2) ->
(Restore target) ->
  C:\Program Files\Mono\lib\mono\msbuild\15.0\bin\NuGet.targets(104,5): error : Unable to load the service index for so
urce https://api.nuget.org/v3/index.json. [D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTo
ols.csproj]
C:\Program Files\Mono\lib\mono\msbuild\15.0\bin\NuGet.targets(104,5): error :   An error occurred while sending the req
uest [D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTools.csproj]
C:\Program Files\Mono\lib\mono\msbuild\15.0\bin\NuGet.targets(104,5): error :   Error: TrustFailure (The authentication
 or decryption has failed.) [D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTools.csproj]
C:\Program Files\Mono\lib\mono\msbuild\15.0\bin\NuGet.targets(104,5): error :   The authentication or decryption has fa
iled. [D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTools.csproj]
C:\Program Files\Mono\lib\mono\msbuild\15.0\bin\NuGet.targets(104,5): error :   Invalid certificate received from serve
r. Error code: 0xffffffff800b0109 [D:\godot-master\build\godot\modules\mono\editor\GodotSharpTools\GodotSharpTools.cspr
oj]

    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:02.26
scons: *** [bin\GodotSharp\Tools\GodotSharpTools.dll] RuntimeError : GodotSharpTools build failed
Traceback (most recent call last):
  File "C:\Python27\Scripts\..\Lib\site-packages\scons-3.0.0\SCons\Action.py", line 1197, in execute
    result = self.execfunction(target=target, source=rsources, env=env)
  File "D:\godot-master\build\godot\modules\mono\SCsub", line 231, in mono_build_solution
    raise RuntimeError('GodotSharpTools build failed')
RuntimeError: GodotSharpTools build failed
   Creating library bin\godot.windows.tools.64.mono.lib and object bin\godot.windows.tools.64.mono.exp
scons: building terminated because of errors.
```

File: modules/mono/editor/GodotSharpTools/Project/ProjectGenerator.cs [hunk 1]
```diff
@@ -6,6 +6,9 @@ namespace GodotSharpTools.Project
 {
     public static class ProjectGenerator
     {
+        const string CoreApiProjectGuid = "{AEBF0036-DA76-4341-B651-A3F2856AB2FA}";
+        const string EditorApiProjectGuid = "{8FBEC238-D944-4074-8548-B3B524305905}";
```
Comment (akien-mga) [inline modules/mono/editor/GodotSharpTools/Project/ProjectGenerator.cs]: Will those always stay the same? Might be worth documenting where those IDs come from/what they mean with a short comment.

**Comment by akien-mga** [comment]
> This PR adds a tiny dependency for the editor (https://github.com/dazinator/DotNet.Glob), and I'm still trying to figure out where the license info should go.

I don't see the dependency added in the `thirdparty` folder, is it downloaded at runtime? Usually the process should be to put the source files in `thirdparty`, document their provenance and license in `thirdparty/README.md` and add licensing details in `COPYRIGHT.txt`.

But if it's a dependency downloaded as a binary blob at compile time, I'm not sure how to handle that (apart from saying "Eeew" because I hate buildsystems that rely on downloading stuff from the Internet :P). I guess we should design a way for the Mono module to register new thirdparty components to add to the Engine singleton's license details (currently parsed from `COPYRIGHT.txt` at compile time).

File: modules/mono/editor/GodotSharpTools/Project/ProjectGenerator.cs (hunk 1 above)
Comment (neikeq) [reply modules/mono/editor/GodotSharpTools/Project/ProjectGenerator.cs]: These are randomly generated unique IDs. They will stay the same, unlike before when we were generating them every time for the same project.

**Comment by neikeq** [comment]
Yes, it's downloaded at build time. The NuGet reference is here: https://github.com/godotengine/godot/pull/23162/files#diff-15a6ab61c9cc330d83e32426b12e15c6R34

**Comment by exts** [comment]
[NSTest2.zip](https://github.com/godotengine/godot/files/2502677/NSTest2.zip)

Works flawlessly, awesome :)

**Comment by exts** [comment]
@neikeq#7308 

```
mono_gc_handle.cpp
modules\mono\csharp_script.cpp(835): error C2065: 'DEBUG_ENABLED': undeclared identifier
modules\mono\csharp_script.cpp(843): error C3861: 'read_all_file_utf8': identifier not found
modules\mono\csharp_script.cpp(2608): error C3861: 'read_all_file_utf8': identifier not found
scons: *** [modules\mono\csharp_script.windows.opt.64.obj] Error 2
scons: building terminated because of errors.
```

`scons p=windows target=release tools=no module_mono_enabled=yes -j3`

When building export templates

**Comment by neikeq** [comment]
@exts Fixed

**Comment by aaronfranke** [comment]
Previous commit (dbaa22329761ce7a4d8bab291871b5b5ba359cbe) builds fine.

After this commit (d47cec43f2f4ef2cd6d1a0acf6e99a9dd8d31eef):

```
scons: *** [bin/GodotSharp/Tools/GodotSharpTools.dll] RuntimeError : Cannot find NuGet executable
Traceback (most recent call last):
  File "/usr/lib/scons/SCons/Action.py", line 1197, in execute
    result = self.execfunction(target=target, source=rsources, env=env)
  File "/home/aaronfranke/workspace/godot/modules/mono/SCsub", line 291, in mono_build_solution
    raise RuntimeError('Cannot find NuGet executable')
```

**Comment by aaronfranke** [comment]
When I compile, it generates several DLL files which are not in the repo and are not in `.gitignore`.

**Comment by KellyThomas** [comment]
@hedgehog90 If you don't have certs configured yet nuget you will need to do something like this:  
https://stackoverflow.com/questions/4926676/mono-https-webrequest-fails-with-the-authentication-or-decryption-has-failed/#49799509

**Comment by hedgehog90** [comment]
@KellyThomas Thanks! That did the trick.

**Comment by neikeq** [comment]
Docs updated: https://github.com/godotengine/godot-docs/pull/1911

**Comment by aaronfranke** [comment]
Exporting doesn't work for me with 3.1 Alpha 2:

```
$ ERROR: connect: Signal 'confirmed' is already connected to given method '_confirmed' in that object.
   At: core/object.cpp:1463.
ERROR: load_from: FATAL: Condition ' !FileAccess::exists(p_path) ' is true.
   At: modules/mono/mono_gd/gd_mono_assembly.cpp:466.
Stacktrace:
```

**Comment by exts** [comment]
@aaronfranke https://github.com/godotengine/godot/commit/9fa1e0b4153102913cddc6380207f5c43e6ebaac make sure you have that fix

**Pull request merged**